File: Connect SDK/Samples C#/RTSP/Start IPC2 RTSP/FormMain.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using System.Runtime.InteropServices;
using Optris.Library.Rtsp;
using System.Collections.Generic;
using System.Net;
using System.Net.NetworkInformation;

namespace IPC2
{
    public partial class FormMain : Form
    {
#if _WIN64
        const Int32 OS = 64;
#else
        const Int32 OS = 32;
#endif
        public static readonly Int32 S_OK = 0;
        public static readonly Int32 S_FALSE = 1;

        private RtspStreamer _rtspStreamer;
        private int _rtspPort;
        private List<IPAddress> ipAddresses;

        public System.Version Version
        {
            get { return System.Reflection.Assembly.GetExecutingAssembly().GetName().Version; }
        }

        Int32 FrameWidth, FrameHeight, FrameDepth, FrameSize, MetadataSize;
        public IPC2 ipc;
        bool ipcInitialized = false, frameInitialized = false;
        Bitmap bmp;
        bool Connected = false;
        bool Colors = false;

        byte[] rgbValues;
        Int16[] Values;

        bool Painted = false;

        public FormMain()
        {
            InitializeComponent();
        }

        private void FormMain_Load(object sender, EventArgs e)
        {
            Text += String.Format(" (Rel. {0} (x{1}))", Version, OS);

            Init(160, 120, 2);
            ipc = new IPC2(1);
            timer2.Enabled = true;
            fillIPListBox();
        }

        private void FormMain_FormClosing(object sender, FormClosingEventArgs e)
        {
            ReleaseIPC();
            releaseStreamer();
        }

        void InitIPC()
        {
            Int64 hr;
            if ((ipc != null) && !ipcInitialized)
            {
                hr = IPC2.Init(0, textBoxInstanceName.Text);

                if (hr != S_OK)
                {
                    ipcInitialized = frameInitialized = false;
                }
                else
                {
                    ipc.OnServerStopped = new IPC2.delOnServerStopped(OnServerStopped);
                    IPC2.SetCallback_OnServerStopped(0, ipc.OnServerStopped);

                    ipc.OnFrameIRInit = new IPC2.delOnFrameInit(OnFrameInit);
                    Int32 u = IPC2.SetCallback_OnFrameInit(0, ipc.OnFrameIRInit);

                    ipc.OnNewFrameIREx2 = new IPC2.delOnNewFrameEx2(OnNewFrameEx2);
                    IPC2.SetCallback_OnNewFrameEx2(0, ipc.OnNewFrameIREx2);

                    ipc.OnInitCompleted = new IPC2.delOnInitCompleted(OnInitCompleted);
                    IPC2.SetCallback_OnInitCompleted(0, ipc.OnInitCompleted);

                    hr = IPC2.Run(0);
                    ipcInitialized = (hr == S_OK);
                }
                label1.Text = (hr != S_OK) ? "NOT CONNECTED" : "OK";

            }
        }

        private void ReleaseIPC()
        {
            Connected = false;
            if ((ipc != null) && ipcInitialized)
            {
                IPC2.Release(0);
                ipcInitialized = false;
            }
        }

        byte LoByte(Int16 val) { return BitConverter.GetBytes(val)[0]; }
        byte HiByte(Int16 val) { return BitConverter.GetBytes(val)[1]; }
        byte clip(Int32 val) { return (byte)((val <= 255) ? ((val > 0) ? val : 0) : 255); }


        void GetBitmap(Bitmap Bmp, Int16[] values)
        {
            Int32 stride_diff;
            // Lock the bitmap's bits.  
            System.Drawing.Rectangle rect = new System.Drawing.Rectangle(0, 0, Bmp.Width, Bmp.Height);
            System.Drawing.Imaging.BitmapData bmpData = Bmp.LockBits(rect, System.Drawing.Imaging.ImageLockMode.ReadWrite, Bmp.PixelFormat);
            stride_diff = bmpData.Stride - FrameWidth * 3;

            // Get the address of the first line.
            IntPtr ptr = bmpData.Scan0;

            if (Colors)
            {
                for (Int32 dst = 0, src = 0, y = 0; y < FrameHeight; y++, dst += stride_diff)
                    for (Int32 x = 0; x < FrameWidth; x++, src++, dst += 3)
                    {
                        Int32 C = (Int32)LoByte(values[src]) - 16;
                        Int32 D = (Int32)HiByte(values[src - (src % 2)]) - 128;
                        Int32 E = (Int32)HiByte(values[src - (src % 2) + 1]) - 128;
                        rgbValues[dst] = clip((298 * C + 516 * D + 128) >> 8);
                        rgbValues[dst + 1] = clip((298 * C - 100 * D - 208 * E + 128) >> 8);
                        rgbValues[dst + 2] = clip((298 * C + 409 * E + 128) >> 8);
                    }
            }
            else
            {
                Int16 mn, mx;
                GetBitmap_Limits(values, out mn, out mx);
                double Fact = 255.0 / (mx - mn);

                for (Int32 dst = 0, src = 0, y = 0; y < FrameHeight; y++, dst += stride_diff)
                    for (Int32 x = 0; x < FrameWidth; x++, src++, dst += 3)
                        rgbValues[dst] = rgbValues[dst + 1] = rgbValues[dst + 2] = (byte)Math.Min(Math.Max((Int32)(Fact * (values[src] - mn)), 0), 255);
            }

            // Copy the RGB values back to the bitmap
            System.Runtime.InteropServices.Marshal.Copy(rgbValues, 0, ptr, rgbValues.Length);

            // Unlock the bits.
            Bmp.UnlockBits(bmpData);
        }

        void GetBitmap_Limits(Int16[] Values, out Int16 min, out Int16 max)
        {
            Int32 y;
            double Sum, Mean, Variance;
            min = Int16.MinValue;
            max = Int16.MaxValue;
            if (Values == null) return;

            Sum = 0;
            for (y = 0; y < FrameSize; y++)
                Sum += Values[y];
            Mean = (double)Sum / FrameSize;
            Sum = 0;
            for (y = 0; y < FrameSize; y++)
                Sum += (Mean - Values[y]) * (Mean - Values[y]);
            Variance = Sum / FrameSize;
            Variance = Math.Sqrt(Variance);
            Variance *= 3;  // 3 Sigma
            min = (Int16)(Mean - Variance);
            max = (Int16)(Mean + Variance);
        }

        void Init(Int32 frameWidth, Int32 frameHeight, Int32 frameDepth)
        {
            FrameWidth = frameWidth;
            FrameHeight = frameHeight;
            FrameSize = FrameWidth * FrameHeight;
            FrameDepth = frameDepth;

            if (_rtspStreamer == null)
            {
                createStreamer();
            }

            timer1.Enabled = true;
            bmp = new Bitmap(FrameWidth, FrameHeight, System.Drawing.Imaging.PixelFormat.Format24bppRgb);
            System.Drawing.Rectangle rect = new System.Drawing.Rectangle(0, 0, bmp.Width, bmp.Height);
            System.Drawing.Imaging.BitmapData bmpData = bmp.LockBits(rect, System.Drawing.Imaging.ImageLockMode.ReadWrite, bmp.PixelFormat);
            Int32 stride = bmpData.Stride;
            bmp.UnlockBits(bmpData);
            rgbValues = new Byte[stride * FrameHeight];
            Values = new Int16[FrameSize];
            pictureBox.Size = new System.Drawing.Size(FrameWidth, FrameHeight);
            UpdateSize();
            frameInitialized = true;
        }

        void UpdateSize()
        {
            Size = new System.Drawing.Size(pictureBox.Right + 20, Math.Max(labelLink.Bottom, pictureBox.Bottom) + 50);
        }

        Int32 OnServerStopped(Int32 reason)
        {
            ReleaseIPC();
            Graphics g = Graphics.FromImage(bmp);
            g.FillRectangle(new SolidBrush(Color.Black), 0, 0, bmp.Width, bmp.Height);
            pictureBox.Invalidate();
            return 0;
        }

        Int32 OnFrameInit(Int32 frameWidth, Int32 frameHeight, Int32 frameDepth)
        {
            Init(frameWidth, frameHeight, frameDepth);
            return 0;
        }

        // will work with Imager.exe release > 3.0 only:
        Int32 OnNewFrameEx2(IntPtr data, IntPtr Metadata)
        {
            if (!frameInitialized)
                return S_FALSE;
            return NewFrame(data, (IPC2.FrameMetadata2)Marshal.PtrToStructure(Metadata, typeof(IPC2.FrameMetadata2)));
        }

        Int32 NewFrame(IntPtr data, IPC2.FrameMetadata2 Metadata)
        {
            labelFrameCounter.Text = "Frame counter: " + Metadata.CounterHW.ToString();

            switch (Metadata.FlagState)
            {
                case IPC2.FlagState.FlagOpen: labelFlag.Text = "open"; labelFlag.ForeColor = Color.Green; labelFlag.BackColor = labelFlag1.BackColor; break;
                case IPC2.FlagState.FlagClose: labelFlag.Text = "closed"; labelFlag.ForeColor = Color.White; labelFlag.BackColor = Color.Red; break;
                case IPC2.FlagState.FlagOpening: labelFlag.Text = "opening"; labelFlag.ForeColor = SystemColors.WindowText; labelFlag.BackColor = Color.Yellow; break;
                case IPC2.FlagState.FlagClosing: labelFlag.Text = "closing"; labelFlag.ForeColor = SystemColors.WindowText; labelFlag.BackColor = Color.Yellow; break;
                default: labelFlag.Text = ""; labelFlag.ForeColor = labelFlag1.ForeColor; labelFlag.BackColor = labelFlag1.BackColor; break;
            }

            for (Int32 x = 0; x < FrameSize; x++)
                Values[x] = Marshal.ReadInt16(data, x * 2);
            if (!Painted)
            {
                GetBitmap(bmp, Values);
                pictureBox.Invalidate();
                Painted = true;
            }

            if(_rtspStreamer != null)
                _rtspStreamer.FeedFrame((Bitmap)bmp.Clone());

            return 0;
        }

        private void btnStream_Click(object sender, EventArgs e)
        {
            if ((_rtspStreamer != null) && (!_rtspStreamer.IsRunning))
            {   // check if setting is changed:
                if (_rtspPort != (int)numericUpDownPort.Value)
                {
                    releaseStreamer();
                    createStreamer();
                }
            }

            if (_rtspStreamer != null)
            {
                try
                {
                    if (_rtspStreamer.IsRunning)
                    {
                        stopStreamer();
                    }
                    else
                    {
                        startStreamer();
                    }
                }
                catch (Exception ex)
                {
                    lblStatus.Text = ex.Message;
                    lblStatus.ForeColor = Color.Red;
                }
            }
        }

        Int32 OnInitCompleted()
        {
            label1.Text = "Connected with #" + IPC2.GetSerialNumber(0);
            Colors = ((TIPCMode)IPC2.GetIPCMode(0) == TIPCMode.Colors);
            Connected = true;
            UpdateSize();
            return S_OK;
        }

        private void timer2_Tick(object sender, EventArgs e)
        {
            if (!ipcInitialized || !Connected) InitIPC();
        }

        private void timer1_Tick(object sender, EventArgs e)
        {
            Painted = false;
        }

        private void pictureBox_Paint(object sender, PaintEventArgs e)
        {
            e.Graphics.DrawImage(bmp, 0, 0);
        }

        private void createStreamer()
        {
            try
            {
                _rtspPort = (int)numericUpDownPort.Value;
                _rtspStreamer = new RtspStreamer(new RtspStreamerOptions()
                {
                    CameraName = "IPC Example",
                    Port = _rtspPort,
                    MinUdpPortPair = 50000,
                    MaxUdpPortPair = 51000,
                    PacketMtu = 1400,
                    FpsNominal = 30
                });

                lblStatus.Text = "Streamer stopped";
                lblStatus.ForeColor = Color.Blue;
            }
            catch (Exception ex)
            {
                lblStatus.Text = ex.Message;
                lblStatus.ForeColor = Color.Red;
            }
        }

        private void releaseStreamer()
        {
            if(_rtspStreamer != null)
                _rtspStreamer.Stop();
            _rtspStreamer = null;
        }

        private void startStreamer()
        {
            btnStream.Text = "Stop stream";
            _rtspStreamer.Start();

            lblStatus.Text = "Streamer running";
            lblStatus.ForeColor = Color.Blue;

            numericUpDownPort.Enabled = false;
        }

        private void stopStreamer()
        {
            btnStream.Text = "Start stream";
            _rtspStreamer.Stop();

            lblStatus.Text = "Streamer stopped";
            lblStatus.ForeColor = Color.Blue;

            numericUpDownPort.Enabled = true;  
        }

        void fillIPListBox()
        {
            ipAddresses = new List<IPAddress>();
            ipAddresses.Add(IPAddress.Parse("127.0.0.1"));
            List<String> adapterplusipaddresses = GetAdapterPlusIPAddresses(ipAddresses);
            
            listBoxIP.Items.Clear();
            listBoxIP.Items.Add("Localhost (" + ipAddresses[0].ToString() + ")");
            foreach(String s in adapterplusipaddresses)
            {
                listBoxIP.Items.Add(s);
            }
            updateLink();
        }

        List<String> GetAdapterPlusIPAddresses(List<IPAddress> ipAddress)
        {
            NetworkInterface[] networkInterfaces = NetworkInterface.GetAllNetworkInterfaces();
		    List<String> result = new List<String>();
		    for (int i = 0; i<networkInterfaces.Length; i++)
            {

			    if ((networkInterfaces[i].NetworkInterfaceType == NetworkInterfaceType.Wireless80211
				    || networkInterfaces[i].NetworkInterfaceType == NetworkInterfaceType.Ethernet
				    || networkInterfaces[i].NetworkInterfaceType == NetworkInterfaceType.GigabitEthernet)
				    )
                {
				    try 
                    {
					    foreach(UnicastIPAddressInformation uipai in networkInterfaces[i].GetIPProperties().UnicastAddresses)
                        {
                            if (uipai.Address.AddressFamily.Equals(System.Net.Sockets.AddressFamily.InterNetwork))
                            {
                                String s = networkInterfaces[i].OperationalStatus == OperationalStatus.Up
                                    ? networkInterfaces[i].Description + " (" + uipai.Address.ToString() + ")"
                                    : networkInterfaces[i].Description + " ( Adapter not connected )";

                                result.Add(s);
                                ipAddress.Add(uipai.Address);
                                break;
                            }
                        }
                    }
				    catch (Exception ex)
				    {
					    break;
				    }
			    }
		    }
		    return result;
	    }

        private void listBoxIP_Click(object sender, EventArgs e)
        {
            updateLink();
        }

        private void numericUpDownPort_ValueChanged(object sender, EventArgs e)
        {
            updateLink();
        }

        private void updateLink()
        {
            int idx = Math.Max(listBoxIP.SelectedIndex, 0);
            String ip = ipAddresses[idx].ToString();
            String port = numericUpDownPort.Value.ToString();
            labelLinkRTSP.Text = "rtsp://" + ip + ":" + port;
        }

        private void labelLinkRTSP_Click(object sender, EventArgs e)
        {
            System.Windows.Forms.Clipboard.SetText(labelLinkRTSP.Text);
        }

    }
}

File: Connect SDK/Samples VC++ .net/Start IPC2/FormMain.cpp
#include "stdafx.h"
#include "FormMain.h"

//#define POLLING

#ifdef _WIN64
#define OS 64
#else
#define OS 32
#endif

namespace IPCSample {

unsigned char clip(int val) { return (val <= 255) ? ((val > 0) ? val : 0) : 255; };

FormMain::FormMain(HWND hostHandle)
{
	InitializeComponent();
	FrameWidth = 0;
	FrameHeight = 0;
	FrameDepth = 0;
    MetadataSize = 0;
    bmp = gcnew Bitmap(100, 100, System::Drawing::Imaging::PixelFormat::Format24bppRgb);
	ipc = nullptr;
	ipcInitialized = false;
	frameInitialized = false;
	Connected = false;
	Painted = false;
}

FormMain::~FormMain()
{
	if (components)
		delete components;
}

void FormMain::AppInit(void)
{
	Text += String::Format(" (Rel. {0} (x{1}))", Version, OS);

	Init(160, 120, 2);
	ipc = gcnew IPC(1);
#ifdef POLLING 
	Application::Idle +=  gcnew EventHandler(this, &FormMain::Application_Idle);
#endif
}

System::Void FormMain::textBoxInstanceName_TextChanged(System::Object^  sender, System::EventArgs^  e)
{
	ReleaseIPC();
}

void FormMain::InitIPC() 
{
	HRESULT hr;
	 if(ipc && !ipcInitialized)
	 {
		hr = ipc->Init(0, textBoxInstanceName->Text);
		
		if(FAILED(hr))
		{
			ipcInitialized = frameInitialized = false;
		}
		else
		{
#ifndef POLLING 
			ipc->OnServerStopped = gcnew IPC::delOnServerStopped(this, &FormMain::OnServerStopped );
			ipc->SetCallback_OnServerStopped(0, ipc->OnServerStopped);

			ipc->OnFrameIRInit = gcnew IPC::delOnFrameInit(this, &FormMain::OnFrameInit );
			ipc->SetCallback_OnFrameInit(0, ipc->OnFrameIRInit);

			ipc->OnNewFrameIREx2 = gcnew IPC::delOnNewFrameEx2(this, &FormMain::OnNewFrameEx );
			ipc->SetCallback_OnNewFrameEx2(0, ipc->OnNewFrameIREx2);

			ipc->OnInitCompleted = gcnew IPC::delOnInitCompleted(this, &FormMain::OnInitCompleted );
			ipc->SetCallback_OnInitCompleted(0, ipc->OnInitCompleted);
#endif
			hr = ipc->Run(0);
			ipcInitialized = SUCCEEDED(hr);
		}
		label1->Text = hr ? "NOT CONNECTED" : "OK";

	 }
}

void FormMain::ReleaseIPC(void) 
{
	Connected = false;
	if(ipc && ipcInitialized)
	{
		ipc->Release(0);
		ipcInitialized = false;
	}
}

void FormMain::AppExit(void)
{
#ifdef POLLING 
	Application::Idle -=  gcnew EventHandler(this, &FormMain::Application_Idle);
#endif
	ReleaseIPC();
	delete ipc;
}

void FormMain::GetBitmap(Bitmap^ Bmp, array<short>^values)
{
	int stride_diff;
	// Lock the bitmap's bits.  
	System::Drawing::Rectangle rect = System::Drawing::Rectangle(0, 0, Bmp->Width, Bmp->Height);
	Imaging::BitmapData^ bmpData = Bmp->LockBits( rect, Imaging::ImageLockMode::ReadWrite, Bmp->PixelFormat );
	stride_diff = bmpData->Stride - FrameWidth*3;

	// Get the address of the first line.
	IntPtr ptr = bmpData->Scan0;

	if(Colors)
	{
		for ( int dst=0, src=0, y=0; y < FrameHeight; y++, dst += stride_diff) 
			for ( int x=0; x < FrameWidth; x++, src++, dst+=3 ) 
			{
				int C = LOBYTE(values[src              ]) - 16;
				int D = HIBYTE(values[src - (src%2)    ]) - 128;
				int E = HIBYTE(values[src - (src%2) + 1]) - 128;
				rgbValues[ dst   ] = clip(( 298 * C + 516 * D           + 128) >> 8);
				rgbValues[ dst+1 ] = clip(( 298 * C - 100 * D - 208 * E + 128) >> 8);
				rgbValues[ dst+2 ] = clip(( 298 * C           + 409 * E + 128) >> 8);
			}
	}
	else
	{
		short mn, mx;
		GetBitmap_Limits(values, &mn, &mx);
		double Fact = 255.0 / (mx - mn);

		for ( int dst=0, src=0, y=0; y < FrameHeight; y++, dst += stride_diff) 
			for ( int x=0; x < FrameWidth; x++, src++, dst+=3 ) 
				rgbValues[dst] = rgbValues[dst+1] = rgbValues[dst+2] = min(max((int)(Fact * (values[src] - mn)), 0), 255);
	}

	// Copy the RGB values back to the bitmap
	System::Runtime::InteropServices::Marshal::Copy( rgbValues, 0, ptr, rgbValues->Length );

	// Unlock the bits.
	Bmp->UnlockBits( bmpData );
}


void FormMain::GetBitmap_Limits(array<short>^Values, short *min, short *max)
{
	int y;
	double Sum, Mean, Variance;
	if(!Values) return;

	Sum = 0;
	for (y=0; y < FrameSize; y++ ) 
		Sum += Values[y];
	Mean = (double)Sum / FrameSize;
	Sum = 0;
	for (y=0; y < FrameSize; y++ ) 
		Sum += (Mean - Values[y]) * (Mean - Values[y]);
	Variance = Sum / FrameSize;
	Variance = Math::Sqrt(Variance);
	Variance *= 3;  // 3 Sigma
	*min = short(Mean - Variance);
	*max = short(Mean + Variance);
}


void FormMain::Application_Idle(Object ^sender, EventArgs ^e)
{
#ifdef POLLING 
	if(Connected && frameInitialized)
	{
        int Size = FrameWidth * FrameHeight * FrameDepth;
        void *Buffer = new char[Size];
        FrameMetadata2 *Metadata = (FrameMetadata2*)(new char[MetadataSize]);
        if (ipc->GetFrameQueue(0))
        if (SUCCEEDED(ipc->GetFrame2(0, 0, Buffer, Size, Metadata, MetadataSize)))
            OnNewFrameEx(Buffer, Metadata);
        delete[] Buffer;
        delete[] Metadata;
    }
#endif
}

HRESULT FormMain::MainTimer100ms(void)
{
	Painted = false;
#ifdef POLLING
	if(ipcInitialized)
	{
		WORD State = ipc->GetIPCState(0, true);
		if(State & IPC_EVENT_SERVER_STOPPED)
			OnServerStopped(0);
		if(!Connected && (State & IPC_EVENT_INIT_COMPLETED))
			OnInitCompleted();
		if(State & IPC_EVENT_FRAME_INIT)
		{
            int frameWidth, frameHeight, frameDepth, frameMetadataSize;
            if (SUCCEEDED(ipc->GetFrameConfig(0, &frameWidth, &frameHeight, &frameDepth)))
            {
                Init(frameWidth, frameHeight, frameDepth);
                if (SUCCEEDED(ipc->GetFrameMetadataSize(0, &frameMetadataSize)))
                    MetadataSize = frameMetadataSize;
            }
        }
	}
#endif
    return S_OK;
}

HRESULT FormMain::MainTimer1000ms(void)
{
	if(Connected)
		labelTempTarget->Text	= String::Format("Target-Temp: {0:##0.0}°C", ipc->GetTempTarget(0));
	return S_OK;
}

void FormMain::RenewFlag(void)
{
	buttonFlagRenew->Text = String::Format("Renew ({0})", ipc->RenewFlag(0) ? "Success" : "Failed");
}

void FormMain::Init(int frameWidth, int frameHeight, int frameDepth)
{
	FrameWidth = frameWidth;
	FrameHeight = frameHeight;
	FrameSize = FrameWidth * FrameHeight;
	FrameDepth = frameDepth;
	timer1->Enabled = true;
	bmp = gcnew Bitmap( FrameWidth , FrameHeight , System::Drawing::Imaging::PixelFormat::Format24bppRgb );
	System::Drawing::Rectangle rect = System::Drawing::Rectangle(0, 0, bmp->Width, bmp->Height);
	Imaging::BitmapData^ bmpData = bmp->LockBits( rect, Imaging::ImageLockMode::ReadWrite, bmp->PixelFormat );
	int stride = bmpData->Stride;
	bmp->UnlockBits( bmpData );
	rgbValues = gcnew array<Byte>(stride * FrameHeight);
	Values = gcnew array<short>(FrameSize);
	pictureBox->Size = Drawing::Size(FrameWidth, FrameHeight);
	UpdateSize();
	frameInitialized = true;
}

void FormMain::UpdateSize(void)
{
	this->Size = Drawing::Size(pictureBox->Right + 20, max(buttonFlagRenew->Bottom, pictureBox->Bottom) + 50);
}

HRESULT FormMain::OnServerStopped(int reason)
{
	ReleaseIPC();
	Graphics ^g = Graphics::FromImage(bmp);
	g->FillRectangle(gcnew SolidBrush(Color::Black), 0, 0, bmp->Width, bmp->Height);
	delete g;
	pictureBox->Invalidate();
	return 0;
}

HRESULT FormMain::OnFrameInit(int frameWidth, int frameHeight, int frameDepth)
{
	Init(frameWidth, frameHeight, frameDepth);
	return 0;
} 

// will work with Imager.exe release > 2.0 only:
HRESULT FormMain::OnNewFrameEx(void * pBuffer, FrameMetadata2 *pMetadata)
{
	labelFrameCounter->Text = "Frame counter HW/SW: " + pMetadata->CounterHW.ToString() + "/" + pMetadata->Counter.ToString();
    String ^s = "PIF  ";
    for (int i = 0; i < pMetadata->PIFnDI; i++)
        s += String::Format("  DI{0}:{1}", i + 1, (pMetadata->PIFDI >> i) & 1);
    for (int i = 0; i < pMetadata->PIFnAI; i++)
        s += String::Format("  AI{0}:{1}", i + 1, pMetadata->PIFAI[i]);
    labelPIF->Text = s;

	switch(pMetadata->FlagState)
	{
	case fsFlagOpen: labelFlag->Text = "open"; labelFlag->ForeColor = Color::Green; labelFlag->BackColor = labelFlag1->BackColor; break;
	case fsFlagClose: labelFlag->Text = "closed"; labelFlag->ForeColor = Color::White; labelFlag->BackColor = Color::Red; break;
	case fsFlagOpening: labelFlag->Text = "opening"; labelFlag->ForeColor = SystemColors::WindowText; labelFlag->BackColor = Color::Yellow; break;
	case fsFlagClosing: labelFlag->Text = "closing"; labelFlag->ForeColor = SystemColors::WindowText; labelFlag->BackColor = Color::Yellow; break;
	default: labelFlag->Text = ""; labelFlag->ForeColor = labelFlag1->ForeColor; labelFlag->BackColor = labelFlag1->BackColor;
	}
	return NewFrame((short*)pBuffer, pMetadata->Counter);
}

HRESULT FormMain::OnNewFrame(char * pBuffer, int frameCounter)
{
	return NewFrame((short*)pBuffer, frameCounter);
}

HRESULT FormMain::NewFrame(short *ImgBuf, int frameCounter)
{
	for ( int x = 0; x < FrameSize; x++ ) 
		Values[x] = ImgBuf[x];
	if(!Painted)
	{
		GetBitmap(bmp, Values);
		pictureBox->Invalidate();
		Painted = true;
	}

	return 0;
}

HRESULT FormMain::OnInitCompleted(void)
{
	label1->Text = "Connected with #" + ipc->GetSerialNumber(0);
	Colors = (TIPCMode(ipc->GetIPCMode(0)) == ipcColors);
	Connected = true;
	UpdateSize();
	return S_OK;
}

}
